sae_lens/saes/sae.py

```diff
@@ -0,0 +1,970 @@
+"""Base classes for Sparse Autoencoders (SAEs)."""
+
+import json
+import warnings
+from abc import ABC, abstractmethod
+from contextlib import contextmanager
+from dataclasses import dataclass, field, fields
+from pathlib import Path
+from typing import Any, Callable, Type, TypeVar
+
+import einops
+import torch
+from jaxtyping import Float
+from numpy.typing import NDArray
+from safetensors.torch import save_file
+from torch import nn
+from transformer_lens.hook_points import HookedRootModule, HookPoint
+from typing_extensions import deprecated, overload
+
+from sae_lens import logger
+from sae_lens.config import (
+    DTYPE_MAP,
+    SAE_CFG_FILENAME,
+    SAE_WEIGHTS_FILENAME,
+    SPARSITY_FILENAME,
+    LanguageModelSAERunnerConfig,
+)
+from sae_lens.loading.pretrained_sae_loaders import (
+    NAMED_PRETRAINED_SAE_LOADERS,
+    PretrainedSaeDiskLoader,
+    PretrainedSaeHuggingfaceLoader,
+    get_conversion_loader_name,
+    handle_config_defaulting,
+    sae_lens_disk_loader,
+)
+from sae_lens.loading.pretrained_saes_directory import (
+    get_config_overrides,
+    get_norm_scaling_factor,
+    get_pretrained_saes_directory,
+    get_repo_id_and_folder_name,
+)
+from sae_lens.regsitry import get_sae_class, get_sae_training_class
+
+T = TypeVar("T", bound="SAE")
+
+
+@dataclass
+class SAEConfig:
+    """Base configuration for SAE models."""
+
+    architecture: str
+    d_in: int
+    d_sae: int
+    dtype: str
+    device: str
+    model_name: str
+    hook_name: str
+    hook_layer: int
+    hook_head_index: int | None
+    activation_fn: str
+    activation_fn_kwargs: dict[str, Any]
+    apply_b_dec_to_input: bool
+    finetuning_scaling_factor: bool
+    normalize_activations: str
+    context_size: int
+    dataset_path: str
+    dataset_trust_remote_code: bool
+    sae_lens_training_version: str
+    model_from_pretrained_kwargs: dict[str, Any] = field(default_factory=dict)
+    seqpos_slice: tuple[int, ...] | None = None
+    prepend_bos: bool = False
+    neuronpedia_id: str | None = None
+
```

Comment:
Fixes `make docs-serve error: TypeError: SAEConfig.__init__() missing 1 required positional argument: 'activation_fn_kwargs'`. The issue occurs because `SAEConfig.from_dict()` doesn't provide defaults for missing required fields during documentation generation.

Solution: Add f`ield(default_factory=dict)` to `activation_fn_kwargs` and move it after non-default fields (required by Python dataclass constraints).

```suggestion
    activation_fn_kwargs: dict[str, Any]
    apply_b_dec_to_input: bool
    finetuning_scaling_factor: bool
    normalize_activations: str
    context_size: int
    dataset_path: str
    dataset_trust_remote_code: bool
    sae_lens_training_version: str
    model_from_pretrained_kwargs: dict[str, Any] = field(default_factory=dict)
    seqpos_slice: tuple[int, ...] | None = None
    prepend_bos: bool = False
    neuronpedia_id: str | None = None
```